synthesizer/process/src/finalize.rs

```diff
@@ -187,80 +178,236 @@ fn finalize_transition<N: Network, P: FinalizeStorage<N>>(
 
-    // Initialize a list for finalize operations.
-    let mut finalize_operations = Vec::new();
-
-    // If there is a finalize scope, finalize the function.
-    if let Some((_, finalize)) = stack.get_function(function_name)?.finalize() {
-        // Retrieve the finalize inputs.
-        let inputs = match transition.finalize() {
-            Some(inputs) => inputs,
-            // Ensure the transition contains finalize inputs.
-            None => bail!("The transition is missing inputs for 'finalize'"),
-        };
-
-        // Initialize the registers.
-        let mut registers = FinalizeRegisters::<N>::new(
-            state,
-            *transition.id(),
-            *function_name,
-            stack.get_finalize_types(finalize.name())?.clone(),
-        );
+    // If the last output of the transition is a future, retrieve and finalize it. Otherwise, there are no operations to finalize.
+    let future = match transition.outputs().last().map(|output| output.future()) {
+        Some(Some(future)) => future,
+        _ => return Ok(Vec::new()),
+    };
 
-        // Store the inputs.
-        finalize.inputs().iter().map(|i| i.register()).zip_eq(inputs).try_for_each(|(register, input)| {
-            // Assign the input value to the register.
-            registers.store(stack, register, input.clone())
-        })?;
+    // Check that the program ID and function name of the transition match those in the future.
+    ensure!(
+        future.program_id() == program_id && future.function_name() == function_name,
+        "The program ID and function name of the future do not match the transition"
+    );
 
-        // Initialize a counter for the index of the commands.
-        let mut counter = 0;
+    // Initialize a list for finalize operations.
+    let mut finalize_operations = Vec::new();
 
+    // Initialize a stack of active finalize states.
+    let mut states = Vec::new();
+
+    // Initialize the top-level finalize state.
+    states.push(initialize_finalize_state(state, future, stack, *transition.id())?);
+
+    // While there are active finalize states, finalize them.
+    while let Some(FinalizeState {
+        mut counter,
+        finalize,
+        mut registers,
+        stack,
+        mut call_counter,
+        mut recent_call_locator,
+    }) = states.pop()
+    {
         // Evaluate the commands.
         while counter < finalize.commands().len() {
             // Retrieve the command.
             let command = &finalize.commands()[counter];
             // Finalize the command.
-            let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| match &command {
+            match &command {
                 Command::BranchEq(branch_eq) => {
-                    counter = branch_to(counter, branch_eq, finalize, stack, &registers)?;
-                    Ok(None)
+                    let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
```

Comment:
really not a fan of `AssertUnwindSafe`, but I guess this doesn't deviate from the current state

Reply:
For context, I made this change so that I could handle the change in control flow when calling `await`.